src/EnergyPlus/WaterThermalTanks.cc

```diff
@@ -8081,66 +8088,42 @@ namespace WaterThermalTanks {
 
 	Real64
-	PLRResidualStratifiedTank(
-		Real64 const HPPartLoadRatio, // compressor cycling ratio (1.0 is continuous, 0.0 is off)
-		Array1< Real64 > const & Par // par(1) = HP set point temperature [C]
+	PLRResidualHPWH(
+		Real64 const HPPartLoadRatio,
+		Array1< Real64 > const & Par
 	)
 	{
 		// FUNCTION INFORMATION:
 		//       AUTHOR         B.Griffith,  Richard Raustad
 		//       DATE WRITTEN   Jan 2012
 		//       MODIFIED
-		//       RE-ENGINEERED
+		//       RE-ENGINEERED  Noel Merket, Oct 2015
 
 		// PURPOSE OF THIS FUNCTION:
 		//  Calculates residual function (desired tank temp - actual tank temp)
 		//  HP water heater output depends on the part load ratio which is being varied to zero the residual.
 
 		// METHODOLOGY EMPLOYED:
-		//  Calls CalcWaterThermalTankStratified to get tank temperature at the given part load ratio (source water mass flow rate)
+		//  Calls with CalcWaterThermalTankMixed or CalcWaterThermalTankStratified to get tank temperature at the given part load ratio (source water mass flow rate)
 		//  and calculates the residual as defined above
-
-		// REFERENCES:
-
-		// USE STATEMENTS:
-		// Using/Aliasing
-
-		// Return value
-		Real64 PLRResidualStratifiedTank;
-
-		// Argument array dimensioning
-
-		// Locals
-		// SUBROUTINE ARGUMENT DEFINITIONS:
-		// par(2) = tank mode
-		// par(3) = water heater num
-		// par(4) = FirstHVACIteration
-		// par(5) = MdotWater
-
-		// FUNCTION PARAMETER DEFINITIONS:
-		//  na
-
-		// INTERFACE BLOCK SPECIFICATIONS
-		//  na
-
-		// DERIVED TYPE DEFINITIONS
-		//  na
-
-		// FUNCTION LOCAL VARIABLE DECLARATIONS:
-		int WaterThermalTankNum; // index of water heater
-		Real64 NewTankTemp; // resulting tank temperature [C]
-		bool FirstHVACIteration; // FirstHVACIteration flag
-
-		WaterThermalTankNum = int( Par( 3 ) );
-		WaterThermalTank( WaterThermalTankNum ).Mode = int( Par( 2 ) );
-		WaterThermalTank( WaterThermalTankNum ).SourceMassFlowRate = Par( 5 ) * HPPartLoadRatio;
-		// FirstHVACIteration is a logical, Par is real, so make 1.0=TRUE and 0.0=FALSE
-		FirstHVACIteration = ( Par( 4 ) == 1.0 );
-		CalcWaterThermalTankStratified( WaterThermalTankNum, HPPartLoadRatio );
+		// Par(1) = HP set point temperature [C]
+		// Par(2) = tank mode
+		// Par(3) = water heater num
+		// Par(4) = FirstHVACIteration
+		// Par(5) = MdotWater
+		int const WaterThermalTankNum = int( Par(3) );
 		WaterThermalTankData & Tank = WaterThermalTank( WaterThermalTankNum );
-		NewTankTemp = FindStratifiedTankSensedTemp( Tank );
-		PLRResidualStratifiedTank = Par( 1 ) - NewTankTemp;
-		return PLRResidualStratifiedTank;
+		HeatPumpWaterHeaterData & HPWH = HPWaterHeater( Tank.HeatPumpNum );
+		bool const isVariableSpeed = ( HPWH.NumofSpeed > 0 );
+		Tank.Mode = int( Par(2) );
+		Tank.SourceMassFlowRate = Par(5) * HPPartLoadRatio;
+		if ( isVariableSpeed ) {
+			CalcWaterThermalTank( WaterThermalTankNum );
+		} else {
+			ConvergeSingleSpeedHPWHCoilAndTank( WaterThermalTankNum, HPPartLoadRatio );
```

Comment:
This is important. This makes sure the heat pump and tank are in sync with the current PLR being evaluated.